internal/codegen/golang/go_type.go

```diff
@@ -1,50 +1,92 @@
 package golang
 
 import (
-	"github.com/kyleconroy/sqlc/internal/compiler"
-	"github.com/kyleconroy/sqlc/internal/config"
+	"github.com/kyleconroy/sqlc/internal/pattern"
+	"github.com/kyleconroy/sqlc/internal/plugin"
 )
 
-func goType(r *compiler.Result, col *compiler.Column, settings config.CombinedSettings) string {
+// XXX: These are copied from python codegen.
+func matchString(pat, target string) bool {
+	matcher, err := pattern.MatchCompile(pat)
+	if err != nil {
+		panic(err)
+	}
+	return matcher.MatchString(target)
+}
+
+func matches(o *plugin.Override, n *plugin.Identifier, defaultSchema string) bool {
+	if n == nil {
+		return false
+	}
+
+	schema := n.Schema
+	if n.Schema == "" {
+		schema = defaultSchema
+	}
+
+	if o.Table.Catalog != "" && !matchString(o.Table.Catalog, n.Catalog) {
+		return false
+	}
+
+	if o.Table.Schema == "" && schema != "" {
+		return false
+	}
+
+	if o.Table.Schema != "" && !matchString(o.Table.Schema, schema) {
+		return false
+	}
+
+	if o.Table.Name == "" && n.Name != "" {
+		return false
+	}
+
+	if o.Table.Name != "" && !matchString(o.Table.Name, n.Name) {
+		return false
+	}
+
+	return true
+}
+
+func goType(req *plugin.CodeGenRequest, col *plugin.Column) string {
 	// Check if the column's type has been overridden
-	for _, oride := range settings.Overrides {
-		if oride.GoTypeName == "" {
+	for _, oride := range req.Settings.Overrides {
+		if oride.GoType.TypeName == "" {
 			continue
 		}
-		sameTable := oride.Matches(col.Table, r.Catalog.DefaultSchema)
-		if oride.Column != "" && oride.ColumnName.MatchString(col.Name) && sameTable {
-			return oride.GoTypeName
+		sameTable := matches(oride, col.Table, req.Catalog.DefaultSchema)
+		if oride.Column != "" && matchString(oride.ColumnName, col.Name) && sameTable {
+			return oride.GoType.TypeName
 		}
 	}
-	typ := goInnerType(r, col, settings)
+	typ := goInnerType(req, col)
 	if col.IsArray {
 		return "[]" + typ
 	}
 	return typ
 }
 
-func goInnerType(r *compiler.Result, col *compiler.Column, settings config.CombinedSettings) string {
-	columnType := col.DataType
+func goInnerType(req *plugin.CodeGenRequest, col *plugin.Column) string {
+	columnType := dataType(col.Type)
```

Comment:
oh, one more thought: it might be worth removing `DataType` from the proto. I fell into the pit where it's not actually filled but available on the struct.

I suspect that https://github.com/stephen/sqlc/blame/32f4c968857273582f71d01b72fdf85589548f1d/internal/codegen/python/gen.go#L199 might be wrong? probably wants to also `dataType(col.Type)`

Reply:
Having both is confusing and error-prone. Let's remove it in another PR